### PR TITLE
[x86/Linux] Revise RealCodeHeader and CodeHeader

### LINKED ISSUE #8342: [x86/Linux] build failure on comdelegate.cpp
I encountered the following build errors while building Core CLR for x86/Linux:
```
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3913:61: error: no member named 'GetUnwindInfo' in '_hpCodeHdr'
    DWORD address = RUNTIME_FUNCTION__BeginAddress(pHeader->GetUnwindInfo(0)) + pCodeInfo->GetRelOffset();
                                                   ~~~~~~~  ^
/home/parjong/projects/dotnet/coreclr/src/inc/corcompile.h:88:58: note: expanded from macro 'RUNTIME_FUNCTION__BeginAddress'
#define RUNTIME_FUNCTION__BeginAddress(prf)             (prf)->BeginAddress
                                                         ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3920:55: error: no member named 'GetNumberOfUnwindInfos' in '_hpCodeHdr'
    for (UINT iUnwindInfo = 0; iUnwindInfo < pHeader->GetNumberOfUnwindInfos(); iUnwindInfo++)
                                             ~~~~~~~  ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3922:56: error: no member named 'GetUnwindInfo' in '_hpCodeHdr'
        PTR_RUNTIME_FUNCTION pFunctionEntry = pHeader->GetUnwindInfo(iUnwindInfo);
                                              ~~~~~~~  ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3945:19: error: no member named 'GetNumberOfUnwindInfos' in '_hpCodeHdr'
    _ASSERTE(pCH->GetNumberOfUnwindInfos() >= 1);
             ~~~  ^
/home/parjong/projects/dotnet/coreclr/src/inc/debugmacros.h:54:39: note: expanded from macro '_ASSERTE'
  #define _ASSERTE(expr) _ASSERTE_MSG(expr, #expr)
                                      ^
/home/parjong/projects/dotnet/coreclr/src/inc/debugmacros.h:45:20: note: expanded from macro '_ASSERTE_MSG'
             if (!(expr)) {                                                 \
                   ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3947:64: error: no member named 'GetUnwindInfo' in '_hpCodeHdr'
    DWORD parentBeginRva = RUNTIME_FUNCTION__BeginAddress(pCH->GetUnwindInfo(0));
                                                          ~~~  ^
/home/parjong/projects/dotnet/coreclr/src/inc/corcompile.h:88:58: note: expanded from macro 'RUNTIME_FUNCTION__BeginAddress'
#define RUNTIME_FUNCTION__BeginAddress(prf)             (prf)->BeginAddress
                                                         ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3950:54: error: no member named 'GetNumberOfUnwindInfos' in '_hpCodeHdr'
    for (COUNT_T iUnwindInfo = 1; iUnwindInfo < pCH->GetNumberOfUnwindInfos(); iUnwindInfo++)
                                                ~~~  ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:3952:52: error: no member named 'GetUnwindInfo' in '_hpCodeHdr'
        PTR_RUNTIME_FUNCTION pFunctionEntry = pCH->GetUnwindInfo(iUnwindInfo);
                                              ~~~  ^
/home/parjong/projects/dotnet/coreclr/src/vm/codeman.cpp:4039:54: error: use of undeclared identifier 'RUNTIME_FUNCTION__EndAddress'
                    else if (ControlPc >= baseAddr + RUNTIME_FUNCTION__EndAddress(&functionEntry, baseAddr))
                                                     ^
```

### PR DESCRIPTION
**Pull request by parjong**
This commit fixes #8342.

### CODE REVIEW

**Comment by parjong** [comment]
\CC @seanshpark

File: src/vm/codeman.h [hunk 1]
```diff
@@ -228,6 +228,43 @@ typedef struct _hpCodeHdr
...
+    typedef PTR_RealCodeHeader    SelfType;
+#else
+    typedef struct _hpCodeHdr     ImplType;
+    typedef PTR_CodeHeader        SelfType;
+#endif
+
+    SelfType Self(void)
+    {
+#ifdef DACCESS_COMPILE
+        return (TADDR) this;
```
Comment (jkotas) [inline src/vm/codeman.h]: This cast is not right for `DACCESS_COMPILE`.

The rules for `DACCESS_COMPILE` are explained in https://github.com/dotnet/coreclr/blob/master/src/inc/daccess.h#L138, and the whole story is in https://github.com/dotnet/coreclr/blob/master/Documentation/botr/dac-notes.md.

**Comment by jkotas** [comment]
As discussed in https://github.com/dotnet/coreclr/pull/8296, it may be better to change strategy and get Linux x86 to compile and work with `WIN64EXCEPTIONS` undefined first.

**Comment by parjong** [comment]
Thanks you for comment. I'll try that approach, too.

**Comment by jkotas** [comment]
Let's close this for now and see where we get with !WIN64EXCEPTIONS.

**Comment by parjong** [comment]
This PR is merged into #8629.

**Pull request closed without merging**